Backpack-SwiftUI/Button/Classes/BPKButton.swift

```diff
@@ -0,0 +1,167 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import SwiftUI
+import Backpack_Common
+
+public struct BPKButton: View {
+    private let title: String
+    private let icon: Icon?
+    private let size: BPKButton.Size
+    private var style: BPKButton.Style = .primary
+    private let action: () -> Void
+    
+    @Binding private var loading: Bool
+    @Binding private var enabled: Bool
+    
+    public init(
+        _ title: String = "",
+        icon: Icon? = nil,
+        loading: Binding<Bool> = .constant(false),
+        enabled: Binding<Bool> = .constant(true),
+        size: BPKButton.Size = .default,
+        action: @escaping () -> Void
+    ) {
+        self.title = title
+        self.icon = icon
+        self._loading = loading
+        self._enabled = enabled
+        self.action = action
+        self.size = size
+    }
+    
+    public var body: some View {
+        Button(action: action) {
+            ZStack {
+                ButtonLoadingContentView(loading: loading, size: size)
+                ButtonContentView(title: title, size: size, icon: icon)
+                    .opacity(loading ? 0 : 1)
+            }
+        }
+        .buttonStyle(buttonStyle)
+        .disabled(!enabled || loading)
+        .clipShape(RoundedRectangle(cornerRadius: .sm))
+    }
+    
+    public func buttonStyle(_ style: BPKButton.Style) -> BPKButton {
+        var result = self
+        result.style = style
+        return result
+    }
+    
+    private var buttonStyle: CurrentStateButtonStyle {
+        CurrentStateButtonStyle(
+            style: style,
+            size: size,
+            iconOnly: isIconOnly,
+            getCurrentState: currentState(isPressed:),
+            colorProvider: ButtonColorProvider(
+                colorSetFactory: DefaultButtonColorSetFactory()
+            )
+        )
+    }
+    
+    private func currentState(isPressed: Bool) -> BPKButton.CurrentState {
+        if !enabled { return .disabled }
+        if loading { return .loading }
+        if isPressed { return .highlighted }
+        return .regular
+    }
+    
+    private var isIconOnly: Bool {
```

Comment:
Maybe this should be more explicit, rather than assumed based on the title availability? The API also feels a little weird having to specify leading/trailing for icon only

How does the accessibility text work? For compose we have created separate constructors for the different types to make the accessibility text required for icon only & make the API more explicit

Reply:
I've addressed this, there's now 2 different inits, there's no way now to create a button without content, either text, text with icon or icon only (with accessibilityLabel required if icon only)

Reply:
I guess technically you could still fake it by providing an empty title (and then end up without an accessibility label)?